crates/ide-completion/src/render.rs

```diff
@@ -147,7 +148,40 @@ pub(crate) fn render_field(
         .set_documentation(field.docs(db))
         .set_deprecated(is_deprecated)
         .lookup_by(name);
-    item.insert_text(field_with_receiver(db, receiver.as_ref(), &escaped_name));
+    if ty.is_fn() || ty.is_closure() {
+        let mut builder = TextEdit::builder();
+        // Using TextEdit, insert '(' before the struct name and ')' before the
+        // dot access, then comes the field name and optionally insert function
+        // call parens.
+
+        if let Some(receiver) = &dot_access.receiver {
+            let range = receiver.syntax().text_range();
+            builder.insert(range.start(), "(".to_string());
+            builder.insert(range.end(), ")".to_string());
+        }
+        builder.replace(
+            ctx.source_range(),
+            field_with_receiver(db, receiver.as_ref(), &escaped_name).into(),
+        );
+
+        let is_fn_expected =
+            ctx.completion.expected_type.as_ref().map_or(false, |ty| ty.is_fn() || ty.is_closure());
+
+        // This could be refactored as method of DotAccessKind
+        let is_parens_needed = if let DotAccessKind::Method { has_parens } = dot_access.kind {
+            !has_parens
+        } else {
+            true
+        };
+
+        if !is_fn_expected && is_parens_needed {
```

Comment:
Was confused on why the `!is_fn_expected` condition was needed, but realized that for the following code, we don't add parentheses since it's expecting a callable type:
```rs
struct S { field: fn() }

fn foo() {
    let a: fn() = S { field: || () }.fi$0;
}
```

`is_fn_expected` could probably be renamed to something like `expects_fn_type`, and if this is not already covered by an existing test, it'd be great to add a test for it too.